p2p/src/secret_connection.rs

```diff
@@ -285,170 +334,126 @@ impl<IoHandler: Read + Write + Send + Sync> SecretConnection<IoHandler> {
-
-        let tag = self
-            .send_cipher
-            .encrypt_in_place_detached(
-                GenericArray::from_slice(self.send_nonce.to_bytes()),
-                b"",
-                &mut sealed_frame[..TOTAL_FRAME_SIZE],
-            )
-            .map_err(|_| Error::Crypto)?;
-
-        sealed_frame[TOTAL_FRAME_SIZE..].copy_from_slice(tag.as_slice());
-
-        Ok(())
-    }
-
-    /// Decrypt AEAD authenticated data
-    fn decrypt(&self, ciphertext: &[u8], out: &mut [u8]) -> Result<usize> {
-        if ciphertext.len() < TAG_SIZE {
-            return Err(Error::Crypto).wrap_err_with(|| {
-                format!(
-                    "ciphertext must be at least as long as a MAC tag {}",
-                    TAG_SIZE
-                )
-            });
-        }
-
-        // Split ChaCha20 ciphertext from the Poly1305 tag
-        let (ct, tag) = ciphertext.split_at(ciphertext.len() - TAG_SIZE);
-
-        if out.len() < ct.len() {
-            return Err(Error::Crypto).wrap_err("output buffer is too small");
-        }
-
-        let in_out = &mut out[..ct.len()];
-        in_out.copy_from_slice(ct);
-
-        self.recv_cipher
-            .decrypt_in_place_detached(
-                GenericArray::from_slice(self.recv_nonce.to_bytes()),
-                b"",
-                in_out,
-                tag.into(),
-            )
-            .map_err(|_| Error::Crypto)?;
-
-        Ok(in_out.len())
-    }
 }
 
-impl<IoHandler> Read for SecretConnection<IoHandler>
+impl<IoHandler> SecretConnection<IoHandler>
 where
-    IoHandler: Read + Write + Send + Sync,
+    IoHandler: TryClone,
+    <IoHandler as TryClone>::Error: 'static + std::error::Error + Send + Sync,
 {
-    // CONTRACT: data smaller than DATA_MAX_SIZE is read atomically.
-    fn read(&mut self, data: &mut [u8]) -> io::Result<usize> {
-        if !self.recv_buffer.is_empty() {
-            let n = cmp::min(data.len(), self.recv_buffer.len());
-            data.copy_from_slice(&self.recv_buffer[..n]);
-            let mut leftover_portion = vec![
-                0;
-                self.recv_buffer
-                    .len()
-                    .checked_sub(n)
-                    .expect("leftover calculation failed")
-            ];
-            leftover_portion.clone_from_slice(&self.recv_buffer[n..]);
-            self.recv_buffer = leftover_portion;
-
-            return Ok(n);
-        }
+    /// For secret connections whose underlying I/O layer implements
+    /// [`tendermint_std_ext::TryClone`], this attempts to split such a
+    /// connection into its sending and receiving halves.
+    ///
+    /// This facilitates full-duplex communications when each half is used in
+    /// a separate thread.
+    ///
+    /// ## Errors
+    /// Fails when the `try_clone` operation for the underlying I/O handler
+    /// fails.
+    pub fn try_split(self) -> Result<(Sender<IoHandler>, Receiver<IoHandler>)> {
```

Comment:
Personally I'd just call this `split` unless there's already an analogous `split` method (i.e. `try_*` is helpful for disambiguating a fallible alternative to an existing method)